login_server/player/state/authenticated_state.py

```diff
@@ -127,17 +132,30 @@ def handle_server_join_first_step(self, request):
         else:
             game_server = self.player.login_server.find_server_by_id(server_field.value)
 
+            if game_server.game_setting_mode != self.player.player_settings.game_setting_mode:
```

Comment:
I think this is the exception to the rule. Looks like this and where you send the player its mode is the only place the player's current mode should be accessed.